upgrade_through_versions_test.py

```diff
@@ -970,60 +793,91 @@ def _multidc_schema_create(self):
-    vars()[cls_name] = type(
-        cls_name,
-        (PointToPointUpgradeBase,),
-        {'test_versions': [from_ver_latest_tag, make_branch_str(to_branch)], '__test__': True})
-
-# create test classes for upgrading from HEAD of one branch to HEAD of next.
-for (from_branch, to_branch) in POINT_UPGRADES:
-    cls_name = ('TestUpgrade_from_' + make_branch_str(from_branch) + '_HEAD_to_' + make_branch_str(to_branch) + '_HEAD').replace('-', '_').replace('.', '_')
-    debug('Creating test upgrade class: {}'.format(cls_name))
-    vars()[cls_name] = type(
-        cls_name,
-        (PointToPointUpgradeBase,),
-        {'test_versions': [make_branch_str(from_branch), make_branch_str(to_branch)], '__test__': True})
-
-# create test classes for upgrading from HEAD of one branch, to latest tag of next branch
-for (from_branch, to_branch) in POINT_UPGRADES:
-    cls_name = ('TestUpgrade_from_' + make_branch_str(from_branch) + '_HEAD_to_' + make_branch_str(to_branch) + '_latest_tag').replace('-', '_').replace('.', '_')
-    to_ver_latest_tag = latest_tag_matching(to_branch)
-    # in some cases we might not find a tag (like when the to_branch is trunk)
-    # so these will be skipped.
-    if to_ver_latest_tag is None:
-        continue
-    debug('Creating test upgrade class: {} with end tag of: {} ({})'.format(cls_name, to_ver_latest_tag, sha_for_ref_name(to_ver_latest_tag)))
-
-    vars()[cls_name] = type(
+
+# used by function def immediately below
+created_class_scope = vars()
+
+
+def create_upgrade_class(version_list, protocol_version=DEFAULT_PROTOCOL_VERSION, bootstrap_test=False):
+    """
+    Dynamically creates a test subclass for testing the given versions.
+
+    'version_list' should be a list of versions ccm will recognize.
+    'bootstrap_test' is a boolean, if True bootstrap testing will be included.
+    """
+    if bootstrap_test:
+        parent_classes = (UpgradeTester, BootstrapMixin)
+    else:
+        parent_classes = (UpgradeTester,)
+
+    cls_name = 'TestUpgradeDynamicClass__' + "__".join(version_list).replace(':', '_').replace('-', '').replace('.', '_')
+    debug("Creating test class {}, parent classes: {}".format(cls_name, parent_classes))
+    created_class_scope[cls_name] = type(
         cls_name,
-        (PointToPointUpgradeBase,),
-        {'test_versions': [make_branch_str(from_branch), to_ver_latest_tag], '__test__': True})
+        parent_classes,
+        {'test_versions': version_list, '__test__': True, 'protocol_version': protocol_version}
+    )
+
+# maps protocol version to c* version(s)
+# Versions are tuples of (major_ver, minor_ver)
+# PROTOCOL_PATHS = {
+#     1: [(2, 0), (2, 1), (2, 2)],
+#     2: [(2, 0), (2, 1), (2, 2)],
+#     3: [(2, 1), (2, 2), (3, 0), (3, 1), TRUNK_VER],
+#     4: [(2, 2), (3, 0), (3, 1), TRUNK_VER]
+# }
+
+create_upgrade_class(['git:cassandra-2.0', 'git:cassandra-2.1'], bootstrap_test=True)
```

Comment:
Now that I see how this is used, it might be better have `create_upgrade_class` be a pure function returning a class, leaving attaching the class to the module at the module level:

```python
Upgrade20HEAD21Bootstrap = create_upgrade_class(name='Upgrade20HEAD21Bootstrap',
                                                test_versions=['git:cassandra-2.0', 'git:cassandra-2.1'],
                                                bootstrap_test=True)
```

You could implement the existing `create_upgrade_class` on top of that, so if you want, e.g., automatic class name generation and/or automatic insertion into `globals()`, you could write that on top of a function implementing the pure-function interface. At the very least, this makes the code a little more decomposed and reusable.